File: app/src/main/java/openfoodfacts/github/scrachx/openfood/utils/NumberParserUtils.java
package openfoodfacts.github.scrachx.openfood.utils;

import androidx.annotation.Nullable;

import org.apache.commons.lang.StringUtils;

import java.util.Map;

public class NumberParserUtils {
    private NumberParserUtils() {

    }

    public static float getAsFloat(@Nullable Object in, float defaultValue) {
        if (in == null) {
            return defaultValue;
        }
        if (in instanceof Number) {
            return ((Number) in).floatValue();
        }
        return (float) parseDouble(in.toString(), defaultValue);
    }

    public static float getAsFloat(@Nullable Map<String, ?> imgDetails, @Nullable String key, float defaultValue) {
        if (imgDetails == null || key == null) {
            return defaultValue;
        }
        return NumberParserUtils.getAsFloat(imgDetails.get(key),defaultValue);
    }

    public static int getAsInt(@Nullable Map<String, ?> imgDetails, @Nullable String key, int defaultValue) {
        if (imgDetails == null || key == null) {
            return defaultValue;
        }
        return NumberParserUtils.getAsInt(imgDetails.get(key),defaultValue);
    }

    public static int getAsInt(@Nullable Object in, int defaultValue) {
        if (in == null) {
            return defaultValue;
        }
        if (in instanceof Number) {
            return ((Number) in).intValue();
        }
        return (int) parseDouble(in.toString(), defaultValue);
    }

    private static double parseDouble(@Nullable String in, double defaultValue) {
        if (StringUtils.isBlank(in)) {
            return defaultValue;
        }
        try {
            return Double.parseDouble(in);
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/utils/NumberParserUtils.kt
package openfoodfacts.github.scrachx.openfood.utils

private fun parseDouble(number: String?, defaultValue: Double) = when {
    number.isNullOrBlank() -> defaultValue
    else -> try {
        number.toDouble()
    } catch (e: NumberFormatException) {
        defaultValue
    }
}

fun getAsInt(number: Any?, defaultValue: Int) = when (number) {
    null -> defaultValue
    is Number -> number.toInt()
    else -> parseDouble(number.toString(), defaultValue.toDouble()).toInt()
}

fun getAsInt(imgDetails: Map<String, *>?, key: String?, defaultValue: Int) = when {
    imgDetails == null || key == null -> defaultValue
    else -> getAsInt(imgDetails[key], defaultValue)
}

fun getAsFloat(imgDetails: Map<String, *>?, key: String?, defaultValue: Float): Float {
    return if (imgDetails == null || key == null) {
        defaultValue
    } else getAsFloat(imgDetails[key], defaultValue)
}

fun getAsFloat(number: Any?, defaultValue: Float) = when (number) {
    null -> defaultValue
    is Number -> number.toFloat()
    else -> parseDouble(number.toString(), defaultValue.toDouble()).toFloat()
}